src/gcode/gcode_to_command.cpp

```diff
@@ -0,0 +1,431 @@
+    }
+    else
+    {
+        spdlog::warn("FeatureType command without feature type index");
+    }
+}
+
+void VisitCommand::update_state(const gcode::ast::InitialTemperatureExtruder& command)
+{
+    if (command.S)
+    {
+        state.extruder_temperatures[command.T ? state.active_tool : command.T.value()] = command.S.value();
+    }
+    else
+    {
+        spdlog::warn("InitialTemperatureExtruder command without temperature");
+    }
+}
+
+void VisitCommand::update_state(const gcode::ast::InitialTemperatureBuildPlate& command)
+{
+    if (command.S)
+    {
+        state.build_plate_temperature = command.S.value();
+    }
+    else
+    {
+        spdlog::warn("InitialTemperatureBuildPlate command without temperature");
+    }
+}
+
+void VisitCommand::update_state(const gcode::ast::BuildVolumeTemperature& command)
+{
+    if (command.S)
+    {
+        state.chamber_temperature = command.S.value();
+    }
+    else
+    {
+        spdlog::warn("BuildVolumeTemperature command without temperature");
+    }
+}
+
+void VisitCommand::update_state(const gcode::ast::T& command)
+{
+    if (command.S)
+    {
+        state.active_tool = command.S.value();
+    }
+    else
+    {
+        spdlog::warn("Tool change command without tool index");
+    }
+}
+
+void VisitCommand::to_proto_path(const gcode::ast::G0_G1& command)
+{
+    const auto move = std::make_shared<botcmd::Move>();
+
+    double delta_e{};
+    if (! command.E.has_value())
+    {
+        delta_e = 0.0;
+    }
+    else if (state.E_positioning == Positioning::Relative)
+    {
+        delta_e = command.E.value();
+    }
+    else
+    {
+        const auto& prev_state = previous_states[previous_states.size() - 1];
+        const auto last_e = prev_state.E[state.active_tool] + prev_state.origin_e[state.active_tool];
+        const auto curr_e = command.E.value() + state.origin_e[state.active_tool];
+        delta_e = curr_e - last_e;
+    }
+
+    // if position is not specified for an axis, move with a relative delta 0 move
+    move->point = {
+        command.X.has_value() ? (command.X.value() + (state.X_positioning == Positioning::Absolute ? state.origin_x : 0.0)) : 0.0,
+        command.Y.has_value() ? (command.Y.value() + (state.Y_positioning == Positioning::Absolute ? state.origin_y : 0.0)) : 0.0,
+        command.Z.has_value() ? (command.Z.value() + (state.Z_positioning == Positioning::Absolute ? state.origin_z : 0.0)) : 0.0,
+        state.active_tool == 0 && command.E.has_value() ? delta_e : 0.0,
+        state.active_tool == 1 && command.E.has_value() ? delta_e : 0.0,
+    };
+    // gcode is in mm / min, bot cmd uses mm / sec
+    move->feedrate = state.F[state.active_tool] / 60.0;
+    move->is_point_relative = {
+        command.X ? state.X_positioning == Positioning::Relative : true,
+        command.Y ? state.Y_positioning == Positioning::Relative : true,
+        command.Z ? state.Z_positioning == Positioning::Relative : true,
+        true,
+        true,
```

Comment:
This is not technically correct, but was intended. Since makerbot doesn't support "reset extruder" position commands we could only set every extruder command to absolute. When extruder positioning is set to absolute for a whole print commutative errors occur, as the printer can't keep up with the extusion and then over extrudes. As a solution we convert every absolute extruder position to a relative move.

Reply:
Make sense. A small comment would be welcome then :)

Reply:
594eee050fc06c719daf12fb56b5b26c7dcea335